File: prep/implementation/practice_iterations/fast_and_slow_pointers_3/happy_number.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

bool is_happy(int n) {
  auto digit_sum_square = [](int m) {
    int sum = 0;

    while (m) {
      sum += (m % 10) * (m % 10);
      m /= 10;
    }
    return sum;
  };

  int fast = n, slow = n;
  do {
    fast = digit_sum_square(digit_sum_square(fast));
    slow = digit_sum_square(slow);

  } while (fast != slow);

  return slow == 1;
}

// O(logn)/ O(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;
    cout << (is_happy(n) ? "true" : "false") << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/fast_and_slow_pointers_4/HappyNumber.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(logn) / O(1)
public class HappyNumber {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        System.out.println(isHappy(n));
      }
    }
  }

  private static boolean isHappy(int n) {
    int fast = n, slow = n;

    do {
      fast = sumSquareDigits(sumSquareDigits(fast));
      slow = sumSquareDigits(slow);
    } while (fast != slow);

    return slow == 1;
  }

  private static int sumSquareDigits(int n) {
    int sum = 0;

    while (n > 0) {
      sum += (n % 10) * (n % 10);
      n /= 10;
    }

    return sum;
  }
}